File: engine/src/main/java/de/hanno/hpengine/engine/model/StaticModel.java
package de.hanno.hpengine.engine.model;

import com.carrotsearch.hppc.IntArrayList;
import de.hanno.hpengine.engine.transform.Transform;
import de.hanno.hpengine.engine.graphics.buffer.Bufferable;
import org.joml.Vector2f;
import org.joml.Vector3f;

import java.util.ArrayList;

public class StaticModel<T extends Bufferable> extends AbstractModel<T> {

    private ArrayList<Vector3f> vertices = new ArrayList<>();
    private ArrayList<Vector2f> texCoords = new ArrayList<>();
    private ArrayList<Vector3f> normals = new ArrayList<>();

    public StaticModel() {
        super(new ArrayList<>());
    }

    public void addVertex(Vector3f vertex) {
        vertices.add(vertex);
    }

    public void addTexCoords(Vector2f texCoords) {
        this.texCoords.add(texCoords);
    }

    public void addNormal(Vector3f normal) {
        normals.add(normal);
    }

    public ArrayList<Vector3f> getVertices() {
        return vertices;
    }

    public ArrayList<Vector2f> getTexCoords() {
        return texCoords;
    }

    public ArrayList<Vector3f> getNormals() {
        return normals;
    }

    public void init() {
        meshIndices = new IntArrayList[meshes.size()];
        triangleCount = 0;
        for(int i = 0; i < meshes.size(); i++) {
            Mesh mesh = meshes.get(i);
            mesh.init();
            meshIndices[i] = new IntArrayList();
            meshIndices[i].add(mesh.getIndexBufferValuesArray());
            triangleCount += mesh.getTriangleCount();
        }
        super.init();
    }

    public Mesh getMesh(int i) {
        return meshes.get(i);
    }

    public void addMesh(Mesh mesh) {
        meshes.add(mesh);
    }


    @Override
    public Vector3f getCenterWorld(Transform transform) {
        return super.getCenterWorld(transform);
    }

    @Override
    public Vector3f[] getMinMaxWorld(Transform transform) {
        return super.getMinMaxWorld(transform);
    }

    @Override
    public float getBoundingSphereRadius(Transform transform) {
        return super.getBoundingSphereRadius(transform);
    }
}

File: engine/src/main/java/de/hanno/hpengine/engine/model/StaticModel.kt
package de.hanno.hpengine.engine.model

import de.hanno.hpengine.engine.model.material.Material
import de.hanno.hpengine.engine.scene.Vertex
import de.hanno.hpengine.engine.scene.VertexStruct
import de.hanno.hpengine.engine.scene.VertexStructPacked
import de.hanno.hpengine.engine.transform.AABB
import de.hanno.hpengine.engine.transform.AABBData.Companion.getSurroundingAABB
import de.hanno.struct.StructArray
import java.io.File

class StaticModel(override val file: File,
                  meshes: List<StaticMesh>,
                  material: Material = meshes.first().material) : AbstractModel<Vertex>(meshes, material) {

    override val path: String = file.absolutePath
    override val boundingVolume: AABB = calculateBoundingVolume()

    override fun calculateBoundingVolume() = AABB(meshes.map { it.spatial.boundingVolume.localAABB }.getSurroundingAABB())

    override val bytesPerVertex = VertexStruct.sizeInBytes

    override val verticesStructArray = StructArray(uniqueVertices.size) { VertexStruct() }.apply {
        for (i in uniqueVertices.indices) {
            val vertex = uniqueVertices[i]
            val (position, texCoord, normal) = vertex
            val target = getAtIndex(i)
            target.position.set(position)
            target.texCoord.set(texCoord)
            target.normal.set(normal)
        }
    }

    override val verticesStructArrayPacked = StructArray(meshes.sumBy { it.vertices.size }) { VertexStructPacked() }.apply {
        var counter = 0
        for(mesh in meshes) {
            for(vertex in mesh.vertices) {
                val (position, texCoord, normal) = vertex
                val target = getAtIndex(counter)
                target.position.set(position)
                target.texCoord.set(texCoord)
                target.normal.set(normal)
                counter++
            }
        }
    }

    override fun toString(): String = "StaticModel($path)"
}
